sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala

```diff
@@ -108,18 +108,22 @@ object IntervalUtils {
 
   def castStringToYMInterval(
       input: UTF8String,
-      // TODO(SPARK-35768): Take into account year-month interval fields in cast
       startField: Byte,
       endField: Byte): Int = {
+
+    def truncatedMonth(month: String) : String = {
+      if (endField == YearMonthIntervalType.YEAR) "0" else month
+    }
+
     input.trimAll().toString match {
-      case yearMonthRegex("-", year, month) => toYMInterval(year, month, -1)
-      case yearMonthRegex(_, year, month) => toYMInterval(year, month, 1)
+      case yearMonthRegex("-", year, month) => toYMInterval(year, truncatedMonth(month), -1)
+      case yearMonthRegex(_, year, month) => toYMInterval(year, truncatedMonth(month), 1)
       case yearMonthLiteralRegex(firstSign, secondSign, year, month) =>
```

Comment:
The regex
```scala
  private val yearMonthLiteralRegex =
    (s"(?i)^INTERVAL\\s+([+|-])?'$yearMonthPatternString'\\s+YEAR\\s+TO\\s+MONTH$$").r
```
doesn't support other year-month interval types, for instance `INTERVAL MONTH`. @cloud-fan Again, this is a duplicate of SQL parser logic.

Reply:
Let's unify them later

Reply:
@AngersZhuuuu Could you open JIRA ticket for this, please.

Reply:
> @AngersZhuuuu Could you open JIRA ticket for this, please.

Sure and will work for ths.